src/Vault.sol

```diff
@@ -0,0 +1,166 @@
+    uint public immutable decimals;
+    
+    constructor(address dai_, address weth_, address priceFeedAddress, uint bufferNumerator_, uint bufferDenominator_) {
+        collateral = IERC20(weth_);
+        debt = IERC20(dai_);
+
+        priceFeed = AggregatorV3Interface(priceFeedAddress);
+        decimals = priceFeed.decimals();
+
+        // collateralization level
+        bufferNumerator = bufferNumerator_;
+        bufferDenominator = bufferDenominator_;
+    }
+
+    ///@dev Users deposit collateral asset into Vault
+    ///@param collateralAmount Amount of collateral to deposit
+    function deposit(uint collateralAmount) external {       
+        deposits[msg.sender] += collateralAmount;
+        bool sent = collateral.transferFrom(msg.sender, address(this), collateralAmount);
+        require(sent, "Deposit failed!");  
+        emit Deposit(address(collateral), msg.sender, collateralAmount);
+    }
+    
+    ///@notice Users borrow debt asset calculated based on collateralization level and their deposits 
+    ///@dev See getMaxDebt() for margined colleteralization colculation
+    ///@param debtAmount Amount of debt asset to borrow
+    function borrow(uint debtAmount) external {             
+        uint maxDebt = getMaxDebt(msg.sender);
+        require(debtAmount <= maxDebt, "Insufficient collateral!");
+
+        debts[msg.sender] += debtAmount;
+        bool sent = debt.transfer(msg.sender, debtAmount);
+        require(sent, "Borrow failed!");       
+        emit Borrow(address(debt), msg.sender, debtAmount);
+    }
+
+    ///@notice Get maximum debt asset borrowable, calculated based on collateralization level
+    ///@dev Collateralization level: bufferNumerator/bufferDenominator (e.g. 8/10 -> 80%) | Think of this as margining
+    ///@param user User address to calculate maximum debt possible based on collateralization level
+    function getMaxDebt(address user) public view returns(uint) {
+        uint availableCollateral = (deposits[user]/bufferDenominator)*bufferNumerator;
+        uint maxDebt = (availableCollateral*10**decimals / get_daiETH_Price());
+        return maxDebt;
+    }
+
+    ///@notice Users repay their debt; in debt asset terms
+    ///@dev This covers partial and full repayment scenarioes
+    ///@param debtAmount Amount of debt asset to repay
+    function repay(uint debtAmount) external {
+        debts[msg.sender] -= debtAmount;
+
+        bool sent = debt.transferFrom(msg.sender, address(this), debtAmount);
+        require(sent, "Repayment failed!");       
+        emit Repay(address(debt), msg.sender, debtAmount);   
+    }
+
+
+    ///@notice Users withdraw their deposited collateral
+    ///@dev This covers partial and full withdrawal; checks for spare capacity before initiating withdrawal
+    ///@param collateralAmount Amount of collateral asset to withdraw
+    function withdraw(uint collateralAmount) external {       
```

Comment:
i appreciate u trying to have a descriptive name, but this var name also seems confusing to me.  i think `amount` is fine or else `withdrawAmount`